src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/ConcurrentPipeWriter.cs

```diff
@@ -59,6 +59,15 @@ public ConcurrentPipeWriter(PipeWriter innerPipeWriter, MemoryPool<byte> pool, o
             _sync = sync;
         }
 
+        public void Reset()
```

Comment:
This should throw if it's not completed.

Reply:
How do we know if the ConcurrentPipeWriter has been completed or not?

The underlying pipe will always be reset along side this type when the stream is reset, so it will throw when not complete, but not from this method.

Reply:
> How do we know if the ConcurrentPipeWriter has been completed or not?

For one, `_completeException` should not be null. And it should be nulled out in reset. Additionally everything that is nulled/zeroed out should already be nulled or zeroed out if the ConcurrentPipeWriter was completed.

Reply:
Another thing to be aware of is that the ConcurrentPipeWriter is responsible for completing the inner PipeWriter. Nothing else should be completing the inner PipeWriter.

And if there's a pending flush at the time the ConcurrentPipeWriter is Completed, ConcurrentPipeWriter.Complete() will set the the `_completeException` to a non-null value and return immediately before completing the inner PipeWriter. ConcurrentPipeWriter will then complete the inner PipeWriter later after all the data is flushed through the inner PipeWriter.

Currently the only way to observe that a ConcurrentPipeWriter is done from the outside is to flush it after the app exits, wait for the flush task to complete, *then* call `ConcurrentPipeWriter`.

_currentFlushTcs, _bufferedWritePending and _aborted would also need to be reset.